libethashseal/Ethash.cpp

```diff
@@ -247,3 +247,76 @@ bool Ethash::shouldSeal(Interface*)
 {
     return true;
 }
+
+bool Ethash::isMining() const
+{
+    return m_farm.isMining();
+}
+
+WorkingProgress Ethash::miningProgress() const
+{
+    if (isMining())
+        return m_farm.miningProgress();
+    return WorkingProgress();
+}
+
+u256 Ethash::hashrate() const
+{
+    u256 r = externalHashrate();
+    if (isMining())
+        r += miningProgress().rate();
+    return r;
+}
+
+std::tuple<h256, h256, h256> Ethash::getEthashWork(Client const& _client)
+{
+    // lock the work so a later submission isn't invalidated by processing a transaction elsewhere.
+    // this will be reset as soon as a new block arrives, allowing more transactions to be
+    // processed.
+    bool oldShould = shouldServeWork(_client);
+    m_lastGetWork = chrono::system_clock::now();
+
+    if (!sealEngine()->shouldSeal(this))
+        return std::tuple<h256, h256, h256>();
+
+    // if this request has made us bother to serve work, prep it now.
+    if (!oldShould && shouldServeWork(_client))
+        onPostStateChanged();
+    else
+        // otherwise, set this to true so that it gets prepped next time.
+        m_remoteWorking = true;
+    manuallySetWork(m_sealingInfo);
```

Comment:
This will be start of `Ethash::getWork`. Pass BlockHeader (m_sealingInfo) as a parameter to `SealEngineFace::getWork`